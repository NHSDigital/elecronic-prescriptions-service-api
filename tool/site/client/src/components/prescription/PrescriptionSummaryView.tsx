import * as fhir from "fhir/r4"
import {Button} from "nhsuk-react-components"
import React from "react"

import ButtonList from "../common/buttonList"
import * as utils from "./fragments/utils"
import {
  EditPrescriptionProps,
  PrescriptionSummaryViewLabel,
  PrescriptionSummary,
  PrescriptionLevelDetails
} from "./fragments"


interface PrescriptionSummaryViewProps {
  prescriptionBundle: fhir.Bundle
  handleDownload?: () => Promise<void>
  editorProps?: EditPrescriptionProps
}

const PrescriptionSummaryView = ({
  prescriptionBundle,
  handleDownload,
  editorProps
}: PrescriptionSummaryViewProps) => {
  const prescription = parsePrescriptionBundle(prescriptionBundle)
  const medicationRequests = prescription.medicationRequests
  const prescriptionSummary = utils.createPrescriptionSummary(prescriptionBundle, medicationRequests)
  const prescriptionLevelDetails = utils.createPrescriptionLevelDetails(
    prescriptionBundle,
    medicationRequests[0]
  )

  return (
    <>
      <PrescriptionSummaryViewLabel editorProps={editorProps} />

      {handleDownload &&
        <ButtonList>
          <Button onClick={() => handleDownload()} type={"button"}>Download this Prescription</Button>
        </ButtonList>
      }

      <PrescriptionLevelDetails {...prescriptionLevelDetails} editMode={editorProps.editMode} />
      <PrescriptionSummary {...prescriptionSummary} />
    </>
  )
}

const parsePrescriptionBundle = (bundle: fhir.Bundle) => {
  const resources = bundle.entry.map(e => e.resource)
  const medicationRequests = resources.filter(r => r.resourceType === "MedicationRequest") as Array<fhir.MedicationRequest>
  const communicationRequests = resources.filter(r => r.resourceType === "CommunicationRequest") as Array<fhir.CommunicationRequest>

  const medicationRequest = medicationRequests[0]
  const patient: fhir.Patient = resolveReference(bundle, medicationRequest.subject)
  const requesterPractitionerRole: fhir.PractitionerRole = resolveReference(bundle, medicationRequest.requester)

  return {
    patient,
    requesterPractitionerRole,
    medicationRequests,
    communicationRequests
  }
}

function resolveReference<T extends fhir.FhirResource>(bundle: fhir.Bundle, reference: fhir.Reference) {
  return bundle.entry.find(e => e.fullUrl === reference?.reference)?.resource as T
}

export {
  PrescriptionSummaryView,
  PrescriptionSummaryViewProps
}
