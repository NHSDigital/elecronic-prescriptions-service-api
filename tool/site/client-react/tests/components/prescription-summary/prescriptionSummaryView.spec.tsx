import * as React from "react"
import PrescriptionSummaryView from "../../../src/components/prescription-summary/prescriptionSummaryView"
import {summaryPrescription} from "./props"
import pretty from "pretty"
import {render} from "@testing-library/react"

test("Renders correctly", async () => {
  const component = <PrescriptionSummaryView {...summaryPrescription}/>
  const {container} = render(component)
  expect(pretty(container.innerHTML)).toMatchSnapshot()
})
