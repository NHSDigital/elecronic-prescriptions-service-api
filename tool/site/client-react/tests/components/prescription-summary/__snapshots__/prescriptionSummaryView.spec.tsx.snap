// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
"<h1 class=\\"nhsuk-label-wrapper\\"><label class=\\"nhsuk-label nhsuk-label--xl\\">Prescription Summary</label></h1><label class=\\"nhsuk-label nhsuk-label--m\\">Patient</label>
<dl class=\\"nhsuk-summary-list\\">
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Name</dt>
    <dd class=\\"nhsuk-summary-list__value\\">CORY, ETTA (MISS)</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">NHS Number</dt>
    <dd class=\\"nhsuk-summary-list__value\\">9990548609</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Date of Birth</dt>
    <dd class=\\"nhsuk-summary-list__value\\">01-Jan-1999</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Gender</dt>
    <dd class=\\"nhsuk-summary-list__value\\">Female</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Address</dt>
    <dd class=\\"nhsuk-summary-list__value\\">1 Trevelyan Square<br>Boar Lane<br>Leeds<br>West Yorkshire<br>LS1 6AE</dd>
  </div>
</dl><label class=\\"nhsuk-label nhsuk-label--m\\">Prescriber</label>
<dl class=\\"nhsuk-summary-list\\">
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Name</dt>
    <dd class=\\"nhsuk-summary-list__value\\">EDWARDS, Thomas (DR)</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Telecom</dt>
    <dd class=\\"nhsuk-summary-list__value\\">01234567890</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Organization</dt>
    <dd class=\\"nhsuk-summary-list__value\\">SOMERSET BOWEL CANCER SCREENING CENTRE (A99968)</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Address</dt>
    <dd class=\\"nhsuk-summary-list__value\\">MUSGROVE PARK HOSPITAL<br>TAUNTON<br>TA1 5DA</dd>
  </div>
  <div class=\\"nhsuk-summary-list__row\\">
    <dt class=\\"nhsuk-summary-list__key\\">Trust / CCG</dt>
    <dd class=\\"nhsuk-summary-list__value\\">TAUNTON AND SOMERSET NHS FOUNDATION TRUST (RBA)</dd>
  </div>
</dl>"
`;
